arch/x86/kvm/vmx/pkvm/hyp/iommu.c

```diff
@@ -322,36 +351,114 @@ static bool sync_shadow_context_entry(struct pgt_sync_data *sdata)
+	struct pkvm_ptdev *ptdev;
+	struct pkvm_pgtable_cap cap;
 	bool updated = false;
-	u8 aw;
+	u8 tt, aw;
+	u16 bdf;
 
-	if (sdata->guest_ptep && sdata->shadow_pa) {
-		tmp.hi = guest_ce->hi;
-		tmp.lo = sdata->shadow_pa | (guest_ce->lo & 0xfff);
+	if (ecap_smts(sdata->iommu_ecap)) {
+		if (sdata->guest_ptep && sdata->shadow_pa) {
+			tmp.hi = guest_ce->hi;
+			tmp.lo = sdata->shadow_pa | (guest_ce->lo & 0xfff);
 
-		if (ecap_smts(sdata->iommu_ecap))
 			/* Clear DTE to make sure device TLB is disabled for security */
 			context_sm_clear_dte(&tmp);
-		else {
+		}
+	} else {
+		/*
+		 * In legacy mode, a context entry is a leaf entry responsible for
+		 * configuring the actual address translation for the given ptdev,
+		 * much like a PASID table entry in scalable mode. So the below logic
+		 * is quite similar to the logic in sync_shadow_pasid_table_entry()
+		 * for scalable mode.
+		 */
+		bdf = sdata->vaddr >> LM_DEVFN_SHIFT;
+		ptdev = iommu_find_ptdev(iommu, bdf, 0);
+
+		if (!ptdev) {
+			ptdev = iommu_add_ptdev(iommu, bdf, 0);
+			if (!ptdev)
+				return false;
+		}
+
+		if (!sdata->guest_ptep) {
+			if (context_lm_is_present(shadow_ce)) {
+				pkvm_setup_ptdev_vpgt(ptdev, 0, NULL, NULL, NULL);
+				pkvm_setup_ptdev_did(ptdev, 0);
+				iommu_del_ptdev(iommu, ptdev);
+
+				goto out;
+			}
+			return false;
+		}
+
+		tt = context_lm_get_tt(guest_ce);
+		switch (tt) {
+		case CONTEXT_TT_MULTI_LEVEL:
+		case CONTEXT_TT_DEV_IOTLB:
+			aw = context_lm_get_aw(guest_ce);
+			if (aw != 1 && aw != 2 && aw != 3) {
+				pkvm_err("pkvm: unsupported address width %u\n", aw);
+
+				pkvm_setup_ptdev_vpgt(ptdev, 0, NULL, NULL, NULL);
+				pkvm_setup_ptdev_did(ptdev, 0);
+				goto out;
+			}
+			cap.level = (aw == 1) ? 3 :
+				    (aw == 2) ? 4 : 5;
+			cap.allowed_pgsz = pkvm_hyp->ept_cap.allowed_pgsz;
+			pkvm_setup_ptdev_vpgt(ptdev, context_lm_get_slptr(guest_ce),
+					      &viommu_mm_ops, &ept_ops, &cap);
+
+			break;
+		case CONTEXT_TT_PASS_THROUGH:
 			/*
-			 * Set translation type to CONTEXT_TT_MULTI_LEVEL to ensure using
-			 * 2nd-level translation and to disable device TLB for security.
+			 * When host IOMMU driver is using pass-through mode, pkvm
+			 * IOMMU will actually use the address translation
+			 * (CONTEXT_TT_MULTI_LEVEL) with the primary VM's EPT
+			 * to guarantee the protection.
 			 */
-			context_lm_set_tt(&tmp, CONTEXT_TT_MULTI_LEVEL);
+			break;
+		default:
+			pkvm_err("pkvm: unsupported translation type %u\n", tt);
+
+			pkvm_setup_ptdev_vpgt(ptdev, 0, NULL, NULL, NULL);
+			pkvm_setup_ptdev_did(ptdev, 0);
+			goto out;
+		}
 
+		pkvm_setup_ptdev_did(ptdev, context_lm_get_did(guest_ce));
+
+		if (!is_pgt_ops_ept(ptdev->pgt))
+			return false;
+
+		tmp = *guest_ce;
+
+		/*
+		 * Always set translation type to MULTI_LEVEL to ensure address
+		 * translation and to disable device TLB for security.
+		 */
+		context_lm_set_tt(&tmp, CONTEXT_TT_MULTI_LEVEL);
+
+		if (tt != CONTEXT_TT_PASS_THROUGH) {
```

Comment:
for this case, we actually do not need do not need above ptdev vpgt setup, right? Just set from guest_ce slptr. Maybe add above ptdev vpgt setting in the patch really need it?

Reply:
Right, we don't need it right now, but we will need when we add shadow page tables.

Reply:
what I mean is in corresponding patch, you do not need setup ptdev vpgt, you can add these codes in your patch to support shadow page tables.

Reply:
Actually I've recalled that it's not just for shadow tables. We already need it in my next patch https://github.com/intel-staging/pKVM-IA/pull/17/commits/3f201ad27e15b6ece8738070f61dbe9bf29c2aed for populating pgstate_pgt.
Ok, I can reorganize these patches.

Reply:
Ok, just reorg the patch is fine!